consensus/hotstuff/signature/packager.go

```diff
@@ -1,26 +1,253 @@
+// It returns:
+//  - (sigData, nil) if successfully unpacked the signature data
+//  - (nil, signature.ErrInvalidFormat) if failed to unpack the signature data
 func (p *ConsensusSigPackerImpl) Unpack(blockID flow.Identifier, signerIDs []flow.Identifier, sigData []byte) (*hotstuff.BlockSignatureData, error) {
-	panic("to be implemented")
+	// decode into typed data
+	var data signatureData
+	err := p.encoder.Decode(sigData, &data)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode sig data %s: %w", err, signature.ErrInvalidFormat)
+	}
+
+	// deseralize the compact sig types
+	sigTypes, err := deseralizeFromBytes(data.SigType, len(signerIDs))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deseralize sig types from bytes: %w", err)
+	}
+
+	// read all the possible signer IDs at the given block
+	consensus, err := p.committees.Identities(blockID, filter.HasRole(flow.RoleConsensus))
+	if err != nil {
+		return nil, fmt.Errorf("could not find consensus committees by block id(%v): %w", blockID, err)
+	}
+
+	// lookup is a map from node identifier to node identity
+	// it is used to check the given signerIDs are all valid signers at the given block
+	lookup := consensus.Lookup()
+
+	// read each signer's signerID and sig type from two different slices
+	// group signers by its sig type
+	stakingSigners := make([]flow.Identifier, 0)
+	randomBeaconSigners := make([]flow.Identifier, 0)
+
+	for i, sigType := range sigTypes {
+		signerID := signerIDs[i]
+		_, ok := lookup[signerID]
+		if !ok {
+			return nil, fmt.Errorf("unknown signer ID (%v) at the given block (%v): %w",
+				signerID, blockID, signature.ErrInvalidFormat)
+		}
+
+		if sigType == hotstuff.SigTypeStaking {
+			stakingSigners = append(stakingSigners, signerID)
+		} else if sigType == hotstuff.SigTypeRandomBeacon {
+			randomBeaconSigners = append(randomBeaconSigners, signerID)
+		} else {
+			return nil, fmt.Errorf("unknown sigType %v, %w", sigType, signature.ErrInvalidFormat)
+		}
+	}
+
+	return &hotstuff.BlockSignatureData{
+		StakingSigners:               stakingSigners,
+		RandomBeaconSigners:          randomBeaconSigners,
+		AggregatedStakingSig:         data.AggregatedStakingSig,
+		AggregatedRandomBeaconSig:    data.AggregatedRandomBeaconSig,
+		ReconstructedRandomBeaconSig: data.RandomBeacon,
+	}, nil
+}
+
+// given that a SigType can either be a Staking Sig or a Random Beacon Sig
+// a SigType can be converted into a bool.
+// it returns:
+// - (false, nil) for SigTypeStakingSig
+// - (true, nil) for SigTypeRandomBeaconSig
+// - (false, signature.ErrInvalidFormat) for invalid sig type
+func sigTypeToBool(t hotstuff.SigType) (bool, error) {
+	if t == hotstuff.SigTypeStaking {
+		return false, nil
+	}
+	if t == hotstuff.SigTypeRandomBeacon {
+		return true, nil
+	}
+	return false, signature.ErrInvalidFormat
+}
+
+// seralize the sig types into a compact format
+func seralizeToBytes(sigTypes []hotstuff.SigType) ([]byte, error) {
```

Comment:
added seralizeToBytes and deseralizeFromBytes to compact the bytes that represents sigTypes 

@AlexHentschel @tarakby @durkmurder

Reply:
I would really prefer if we use library for it. I feel like something so low level should be in separate module and well covered by tests. A good candidate https://pkg.go.dev/github.com/dropbox/godropbox/container/bitvector.